DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/test_suite/functional/vlan/test_vlan_priority.py

```diff
@@ -0,0 +1,92 @@
+
+
+async def test_vlan_priority_tag(testbed):
+    """
+    Test Name: VLAN priority
+    Test Suite: suite_functional_vlan
+    Test Overview: Test packet forwarding with VLAN priority
+    Test Procedure:
+    1. Initiate test params.
+    2. Set links to vlans.
+    3. Map receiving and non receiving dut_ports.
+        Port 1 -> tx_port
+        Ports 2, 3, 4 -> rx_ports
+    4. Setup packet stream(s) for the broadcast packet.
+    5. Send traffic to rx_ports.
+    6. Verify traffic on receiving ports.
+    7. Verify priority was not stripped from the packet.
+    """
+
+    # 1. Initiate test params.
+    tgen_dev, dent_devices = await tgen_utils_get_dent_devices_with_tgen(testbed, [], 2)
+    if not tgen_dev or not dent_devices:
+        print("The testbed does not have enough dent with tgn connections")
+        return
+    device = dent_devices[0].host_name
+    tg_ports = tgen_dev.links_dict[device][0]
+    dut_ports = tgen_dev.links_dict[device][1]
+    vlan_priority = 6
+    vlan_id = 5
+
+    port_map = ({"port": 0, "settings": [{"vlan": vlan_id, "untagged": False, "pvid": True}]},
+                {"port": 1, "settings": [{"vlan": vlan_id, "untagged": False, "pvid": True}]})
+
+    await configure_bridge_setup(device, dut_ports)
+
+    # 2. Set links to vlans.
+    await configure_vlan_setup(device, port_map, dut_ports)
+
+    # 3. Map receiving and non receiving dut_ports.
+    dev_groups = tgen_utils_dev_groups_from_config(
+        [{"ixp": tg_ports[0], "ip": "100.1.1.2", "gw": "100.1.1.6", "plen": 24, },
+         {"ixp": tg_ports[1], "ip": "100.1.1.3", "gw": "100.1.1.6", "plen": 24, }])
+    await tgen_utils_traffic_generator_connect(tgen_dev, tg_ports, dut_ports, dev_groups)
+
+    tx_ports = dev_groups[tg_ports[0]][0]["name"]
+    rx_ports = [dev_groups[tg_ports[1]][0]["name"]]
+
+    # 4. Setup packet stream(s).
+    streams = {f"Traffic with VLAN priority": {
+        "type": "raw",
+        "protocol": "802.1Q",
+        "ip_source": tx_ports,
+        "ip_destination": rx_ports,
+        "srcMac": "02:00:00:00:00:01",
+        "dstMac": "02:00:00:00:00:02",
+        "vlanID": vlan_id,
+        "vlanPriority": vlan_priority,
+        "egress_track_by": ["vlan", "pcp"]
+    }}
+
+    await tgen_utils_setup_streams(tgen_dev, config_file_name=None, streams=streams)
+
+    # 5. Send traffic to rx_ports.
+    await tgen_utils_start_traffic(tgen_dev)
+    await asyncio.sleep(5)
+    await tgen_utils_stop_traffic(tgen_dev)
+
+    # 6. Verify traffic.
+    stats = await tgen_utils_get_traffic_stats(tgen_dev, "Flow Statistics")
+    ti_to_rx_port_map = get_traffic_port_vlan_mapping(streams, port_map, tg_ports)
+    for row in stats.Rows:
+        egress_stats = await tgen_utils_get_egress_stats(tgen_dev, row)
+        # 7. Verify priority was not stripped from the packet.
+        assert int(egress_stats.Rows[1]["Egress Tracking 2"]) == vlan_priority, "No/wrong VLAN priority in the packet"
```

Comment:
You can have more than 2 rows, or you can have only 1 row (egress_stats.Rows[0]). I would suggest that you use a loop here:
```python
for row in stats.Rows:
    egress_stats = await tgen_utils_get_egress_stats(tgen_dev, row)
    for idx, e_row in enumerate(egress_stats.Rows):
        if idx == 0:
            continue
        assert int(e_row["Egress Tracking 2"]) == vlan_priority, "No/Wrong VLAN priority in the packet"
```

Reply:
I edited config to have 2 ports only, wanted to eliminate the `for loop`

Reply:
Then you have to check that the Rx count for that vlan priority is correct:
```python
assert int(egress_stats.Rows[1]["Egress Tracking 2"]) == vlan_priority
assert int(egress_stats.Rows[1]["Rx Packets"]) == int(egress_stats.Rows[0]["Tx Packets"])
```
You cannot be sure how your device will behave and how it will forward the traffic back to Ixia. You might have a situation where you have 0% loss, row[1] has the correct pcp, but you also have additional rows (2, 3, ...) with the wrong pcp. You need the loop to check that you don't have any rows that should not be there.